src/main/scala/com/microsoft/hyperspace/Hyperspace.scala

```diff
@@ -165,19 +171,41 @@ class Hyperspace(spark: SparkSession) {
   }
 }
 
-object Hyperspace {
-  private lazy val context = new ThreadLocal[HyperspaceContext]
+object Hyperspace extends Logging {
+  // Maintain HyperspaceContext for each Spark session, since Hyperspace depends on
+  // the session's properties such as configs, etc.
+  // Value as (thread id, hyperspace context) to enforce one client thread per Spark session
+  // as Hyperspace doesn't guarantee thread safety.
+  private lazy val contexts =
+    new ConcurrentHashMap[SparkSession, (Long, HyperspaceContext)]().asScala
 
   private[hyperspace] def getContext(spark: SparkSession): HyperspaceContext = {
-    val ctx = context.get()
-    if (ctx == null || !ctx.spark.equals(spark)) {
-      // Ensure the Spark session the caller provides is the same as
-      // the one HyperspaceContext is using because Hyperspace depends on the
-      // session's properties such as configs, etc.
-      context.set(new HyperspaceContext(spark))
+    val ctx = contexts.get(spark)
+    val threadId = Thread.currentThread().getId
+    if (!ctx.isDefined) {
+      contexts.synchronized {
+        if (contexts.size >= HyperspaceConf.maxNumAllowedConcurrentSparkSessions(spark)) {
+          throw HyperspaceException("Too many Spark sessions are using Hyperspace. " +
+            "Please increase the config `spark.hyperspace.maxNumAllowedConcurrentSparkSessions`. " +
+            s"Current limit: ${contexts.size}")
+        }
+        val newCtx = new HyperspaceContext(spark)
+        contexts.put(spark, (threadId, newCtx))
+        newCtx
+      }
+    } else {
+      if (ctx.get._1 != threadId) {
+        // scalastyle:off
+        println("here")
+        throw HyperspaceException("Hyperspace does not support concurrent threads using one " +
```

Comment:
How about: "Hyperspace does not support multiple threads sharing the same Spark session. Please create a Spark session per thread. Current thread id ..."?